drivers: flash: mspi_nor: support MODE_SINGLE and MODE_QUAD_1_4_4

masz-nordic opened the pull request:
~Depends on https://github.com/zephyrproject-rtos/zephyr/pull/80042~

masz-nordic committed:
drivers: flash: mspi_nor: support MODE_SINGLE and MODE_QUAD_1_4_4

Extend driver to support single lane and 1-4-4 IO modes.
Move flash chip quirks to a separate file.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

swift-tk reviewed (commented):
Which 1-4-4 flash device are we initially targeting now?
Does this device not require switching between MSPI_IO_MODE_SINGLE and MSPI_IO_MODE_QUAD_1_4_4 when accessing register and accessing memory?

masz-nordic commented:
@swift-tk
It's MX25R6435F. I've added switching to single mode when necessary.

swift-tk reviewed (commented):
While it is a good start to use struct flash_mspi_nor_cmd and flash_mspi_nor_cmds to normalize differences between flash devices, we should take one step further by excluding the device specific coding out of the genertic jedec nor driver. Additionally, these flash_mspi_nor_cmd and flash_mspi_nor_cmds can not handle difference in dummies and even address length when the size > 8MB. 

I have been looking at the spi-nor in linux https://github.com/torvalds/linux/blob/a7f2e10ecd8f18b83951b0bab47ddaf48f93bf47/drivers/mtd/spi-nor/ and think it is good idea to have a core that only does jedec specified operations and then device/vendor specific file that deal with the differences by fixups hooks.

Failing to assign certain members of these structs before passing it to the controller can cause issues, which is why I included a `struct mspi_dev_cfg devcfg` in the flash data to keep track of all the device-specific settings.

masz-nordic commented:
@swift-tk
>we should take one step further by excluding the device specific coding out of the genertic jedec nor driver

Do you think it should be part of this PR? I am hoping to limit the scope to my use-case (since this driver was MX-specific already).
I know @anangl is working on adding support for configuration based on SFDP.

swift-tk commented:
> @swift-tk
> 
> > we should take one step further by excluding the device specific coding out of the genertic jedec nor driver
> 
> Do you think it should be part of this PR? I am hoping to limit the scope to my use-case (since this driver was MX-specific already). I know @anangl is working on adding support for configuration based on SFDP.

No, it doesn't have to be in this PR... Just mark it as TODO then. It will have to be moved from the generic core one way or another.
SFDP is not the solution here though because it can only take care of what jedec had specified. Although I'm not sure if the optional pages in the SFDP could help on some of differences.

masz-nordic committed:
boards: nordic: update binding for MX25U

Separate compatible is required for Flash MSPI quirks.

Signed-off-by: Marcin Szymczyk <marcin.szymczyk@nordicsemi.no>

swift-tk commented on drivers/flash/flash_mspi_nor.h:
```diff
@@ -0,0 +1,248 @@
+			.cmd = SPI_NOR_OCMD_RD,
+			.cmd_length = 2,
+			.addr_length = 4,
+			.rx_dummy = 20,
```
So what should we do about these hardcoded values? These can change based on SCK frequency.
Maybe create vendor.c and add API to query the necessary info like dummy and address length?

swift-tk reviewed (commented):
The flash_mspi_nor_quirks.h looks more like a mx25r.c rather than a common include...
Suggest to add vendor.c and move mx25r specifics to it.

tomchy committed:
drivers: flash: Optimize mspi_nor driver memory

Move MSPI NOR commands to rodata.
Replace array with empty padding (~1kB) with macro-based assignments.

Ref: NCSDK-32779

Signed-off-by: Tomasz Chyrowicz <tomasz.chyrowicz@nordicsemi.no>

swift-tk commented on drivers/flash/flash_mspi_nor.c:
```diff
@@ -624,8 +774,60 @@ static DEVICE_API(flash, drv_api) = {
+	DT_INST_STRING_TOKEN(inst, quad_enable_requirements))
+#define FLASH_DW15_QER(inst) COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, quad_enable_requirements), \
+	(FLASH_DW15_QER_VAL(inst)), (JESD216_DW15_QER_VAL_NONE))
+
```
Could you align the slashes? It looks way better below than what is above....

swift-tk reviewed (approved):
Looks pretty good!! Thanks @masz-nordic.

danieldegrasse reviewed (commented):
I generally like the direction of this change, but wondering if it makes sense to define a table of `struct flash_mspi_nor_cmds` by JEDEC ID rather than attempting to be generic. I think that values like the `rx_dummy` cycles will end up being device specific.

One example of a way we could do this is how openocd defines SPI flashes, in a table with JEDEC IDs as a key: https://github.com/openocd-org/openocd/blob/master/src/flash/nor/spi.c#L41

Not saying we need to make this change in this PR, but I wonder if this is a reasonable direction to take for this driver longer term

swift-tk commented on drivers/flash/flash_mspi_nor.h:
```diff
@@ -0,0 +1,248 @@
+		.cmd = SPI_NOR_CMD_READ_FAST,
+		.cmd_length = 1,
+		.addr_length = 3,
+		.rx_dummy = 8,
```
Well most cmds are the same though, the ones that we need to deal with are the ones that are not common(or in the jedec spec) which should falls into quirks.

swift-tk replied:
For the dummy cycles, I've suggested add additional API to query in vendor.c with ID and frequency. So basically what linux spi is doing is to catagorize the flash devices under each flash manufacturer(vendor).c files and have the jedec spi core to call fixup(quirk) API to conduct vendor specific operations. For dummy cycles, it can be more general but still vendor specific.

danieldegrasse commented on drivers/flash/flash_mspi_nor.h:
```diff
@@ -0,0 +1,248 @@
+		.cmd = JESD216_CMD_READ_SFDP,
+		.cmd_length = 1,
+		.addr_length = 3,
+		.rx_dummy = 8,
```
Looking at this closer, `rx_dummy` can't be 8 here- the MSPI driver does not support a nonzero `rx_dummy` value when using SINGLE mode, right?

https://github.com/zephyrproject-rtos/zephyr/blob/e1a0350d23706fa1201d5f37190c77141abe8197/drivers/mspi/mspi_dw.c#L983

swift-tk replied:
In jedec spec, that‘s the sfdp default in serial mode, if a controller does not support it, it is their problem.... which should not affect how we write the generic jedec driver.

swift-tk replied:
In jedec spec, that‘s the sfdp default in serial mode, if a controller does not support it, it is their problem.... which should not affect how we write the generic jedec driver.

danieldegrasse replied:
Sure, I'm aware that is how the command is defined in the spec. In the past I've usually seen it implemented by just shifting the read data by 8 bits. This is fine though- I'll log an issue against the `mspi` driver, since this is the driver that will have this issue: https://github.com/zephyrproject-rtos/zephyr/blob/e9fd4b14de02b75ca416a6898118272f04e29407/drivers/mspi/mspi_dw.c#L983
